Updated to path-to-regexp@3.0.0

wesleytodd opened the pull request:
This would probably be a major version bump for router since the path matching semantics are so different.

Also. you will notice a change in the tests.  I opened a separate issue to discuss, #41.

dougwilson commented:
> Also. you will notice a change in the tests. I opened a separate issue to discuss, #41.

The test has been fixed for quite a while on master. I can see this PR cannot even be merged by GitHub, so I think you have it based off some stale code.

wesleytodd commented:
Yep, forgot to update my forked version.  Will do that now, sorry.

dougwilson commented:
And yes, bumping would be 2.0, which is fine :) Can we get a migration guide written up for users to understand how to translate their routes from the old style to the new style to add with this? I think that's really the only blocker.

wesleytodd commented:
Based on the tests there is one breaking use case because `*` is not a valid regexp.  This means `/:foo(*)` needs to become `/:foo(.*)`.

I think @blakeembrey could probably provide a more comprehensive list of changes, but then I would be happy to write them up into a guide.

blakeembrey commented:
I've written the differences a bundle of times, I think the most comprehensive was for Doug in the Express repo.

dougwilson commented:
The tests in this module do not actually exercise the different syntaxes currently, so knowing what broke is not possible from this test suite (for example, I don't see you adding any tests for all the new syntax in 1.x, continuing that trend ;) ).

wesleytodd commented:
I can add a bunch more tests for the supported syntax if you like.  It would seem weird to test features of `path-to-regexp` in this module though, so I would prefer to keep it the way it is.

For documentation, is what is [listed here](https://github.com/pillarjs/path-to-regexp#compatibility-with-express--4x) comprehensive?  Can we just do a link here to that section?

wesleytodd commented:
I cannot find any more docs that what is in that readme.  So I added them here.

As for tests, I think testing dependencies is not a great decision.  But I understand the reasoning here because `path-to-regexp` makes up a HUGE portion of the interface for the router.  If we could come up with a way to exercise the tests from the underlying module where when it updated, we automatically were testing the new features, then I think it would be fine.  Otherwise it is just a maintenance headache.

Either way, this update does not make us worse off.  So IMO it could be merged.

dougwilson commented:
I agree that extensively testing dependencies is not great. But sometimes I still think that at least basic tests of the main features of a dependency _as they are directly exposed through the interface of this module_ helps flush out accidental regressions upstream and can help us understand if we can bump a major in that dep without bumping our dep.

dougwilson commented:
Thanks for starting on the migration guide :) ! One thing I noticed right away is that the migration guides doesn't provide any guidance on whatever the reason the tests were changed in order to up this dep?

wesleytodd commented:
I added tests, I think they should be satisfactory, hits the major points.  

As for the guide thing, I am really unsure what the core changes were, so I don't actually know what to add.  I am happy to add what other people post here, or let them add to this PR as necessary.

gajus commented:
The latest version is 2.0.0.

dougwilson commented:
Yea, so we still need a migration guide from the respect of this module. Users need to understand what routes they need to change and how during the upgrade or they'll never be able to upgrade their apps successfully. It seems the 2.0 release breaks even more tha. The 1.0, which is fine, but I'm not even clear on how to upgrade the route paths for the transition.

blakeembrey commented:
Just a note, but compatibility is still listed in https://github.com/pillarjs/path-to-regexp#compatibility-with-express--4x. The primary difference between 1.x and 2.x was that I needed to drop support for `*` - it wasn't clear how it was meant to be encoded when converting a "path string" like `/*` into `/foo/bar`.

wesleytodd commented:
Ok, I just updated this PR.  But it appears that something also changed with the handling of trailing slashes.  @blakeembrey Know what I need to do to fix the two tests I added `skip` to?

wesleytodd commented on test/route.js:
```diff
@@ -633,23 +633,23 @@ describe('Router', function () {
           request(server)
           .get('/bar')
-          .expect(200, {'0': 'bar', 'foo': 'bar'}, cb)
+          .expect(200, {'foo': 'bar'}, cb)
```
Hmm, I did not do this change.  I wonder what I did.  I will look into why this is changed and fix it before we merge this.

dougwilson replied:
I'm not sure if we need to keep the old behavior if the new behavior is better, but we need to clearly explain how someone looking at their routers will know that the result of req.param is going to change

wesleytodd replied:
Ok, I will add this to the guide and leave the change here.  I do think the named argument is what I would expect from the route defined in this test.

wesleytodd replied:
done

wesleytodd commented:
I just updated the upgrade guide as well.  Any thoughts or changes necessary?

dougwilson commented:
Does this include the changes between 0.1.x and 2.x or just 0.2.x and 2.x? What were the differences between 0.1.x and 0.2.x, again?

wesleytodd commented:
AFAIK, this guide is ALL of the changes since `0.1.7`.

dougwilson commented on README.md:
```diff
@@ -332,6 +332,22 @@ curl http://127.0.0.1:8080/such_path
+The main change is the update to `path-to-regexp@2.0.0`, which has a few breaking changes:
+
+- No longer a direct conversion to a RegExp with sugar on top - it's a path matcher with named and unnamed matching groups
+  - It's unlikely you previously abused this feature, it's rare and you could always use a RegExp instead
```
We should explain how to know if you are using the feature rather than editorial that "you're probably not using it"

wesleytodd replied:
@blakeembrey I copied this from your readme.  Can you give me a short blurb to help explain what this means?

blakeembrey replied:
Sure. It's anywhere someone is abusing stuff like `/\\d+` outside of a matching group (e.g. instead of `/(\\d+)`). It's possible it's used in places today, but wouldn't be too useful since it wouldn't result in a match the user could use.

blakeembrey replied:
If you end up coming up with a better or more succinct explanation, happy to replace the one in the README 😄

dougwilson replied:
You'd be surprised what people do with these routes. Express 4.16.1 is a release just because it seemed to be popular to do something thag shouldn't work, but did.

wesleytodd replied:
Question: What specifically does this mean?

> Parameters have suffixes that augment meaning - `*`, `+` and `?`. E.g. `/:user*`

Couldn't we always use those suffixes?  How are they different from `0.1.7` to `2.0.0`?

blakeembrey replied:
No, 0.x never had a concept of "suffixes". There's places where it happened to work because everything is treated as a regexp with based search and replace. E.g. `/*` became `/(.*)` in search and replace (that's the reason you can't use `*` in Express.js). It also means that, yes, stuff like `/:user?` might have worked but purely because of the search and replace implications and definitely not consistently (for instance, `/:test*` never would have done anything expected).

Note: I haven't tested the specific examples above, but it's possible a comparison (0.x -> 2.x) might be the easiest way to understand the differences.

wesleytodd replied:
Yeah that is probably why those worked for me :)

I was actually thinking it might be nice to update the demo app [1] to support multiple versions, which might be helpful for people to discover the differences for their routes.  Maybe I can work on that next.

[1]: http://forbeslindesay.github.io/express-route-tester/

wesleytodd replied:
Ok, got a PR for that feature in the demo app: https://github.com/ForbesLindesay/express-route-tester/pull/11

wesleytodd replied:
If you all need to see the differences in the versions you can view it here until the PR is merged: https://wesleytodd.github.io/express-route-tester/

dougwilson commented on README.md:
```diff
@@ -332,6 +332,22 @@ curl http://127.0.0.1:8080/such_path
+- No longer a direct conversion to a RegExp with sugar on top - it's a path matcher with named and unnamed matching groups
+  - It's unlikely you previously abused this feature, it's rare and you could always use a RegExp instead
+- All matching RegExp special characters can be used in a matching group. E.g. `:user(*)` needs to become `/:user(.*)`
+  - Other RegExp features are not support - no nested matching groups, non-capturing groups or look aheads
```
Typo in "support"

wesleytodd replied:
fixed

blakeembrey commented on test/router.js:
```diff
@@ -730,7 +730,7 @@ describe('Router', function () {
     })
 
-    it('should match if path has trailing slash', function (done) {
+    it.skip('should match if path has trailing slash', function (done) {
```
Ah, this one is an ongoing discussion but will likely be fixed in a 2.1 release, sorry - https://github.com/pillarjs/path-to-regexp/issues/123. The old one only passed because it always trimmed the trailing slash, the new one is leaving it there but I'm going to adjust the behaviour in non-ending mode to support this match behaviour.

blakeembrey replied:
@dougwilson I could use your input though, does this feel valid to you. Building up a path, from looking at it, should be `/foo//` and `/foo//foo`, but in this mode it acts differently to written. I'm interested in the viewpoint in support of this oddity.

dougwilson replied:
I'll need to get a computer to understand the question, because mobile provides no way to view outside the diff to even see the path and url in the test.

Also thinking on this, i think we should just do a router 2 alpha now. Should we wait for 2.1.0 first or just put this pr back to 1.x and release a router 2 alpha 1 with it and follow with a alpha 2 after the 2.x regexp changes are settled?

wesleytodd replied:
I am available this weekend to finish up what we need to get this merged into a 2 alpha, so if we could do that it would be awesome.  IMO getting this sooner will help people adjust better and give them more time to work with the new ways.

dougwilson replied:
Yea. That's why I'm wondering if we just start with the first alpha using path to regexp 1.x

wesleytodd replied:
@blakeembrey How much is involved in putting together a PR for https://github.com/pillarjs/path-to-regexp/issues/123#issuecomment-335948313?  I haven't actually worked in `path-to-regxp` before, but if its simple enough I can take a stab at it before we decide what to do with this alpha.  If its too much, I can just revert this back to `1.7.0` simple enough.

blakeembrey replied:
It should be very simple and I can implement it now, if that's what you need.

blakeembrey replied:
It's mostly been a matter of figuring out how it _should work_. Right now, based on feedback and these tests, I'm thinking (using `/foo/` input):

* `end: true, strict: false` - `/foo//?`
* `end: true, strict: true` - `/foo/`
* `end: false, strict: false` - `/foo(?=/)`
* `end: false, strict: true` - `/foo/`

blakeembrey replied:
Actually, comparing this one to https://github.com/pillarjs/path-to-regexp/issues/123, it's slightly different. Supporting https://github.com/pillarjs/path-to-regexp/issues/123 makes sense and is pretty straightforward. The reasoning behind supporting this one is less so. I'll wait for @dougwilson for feedback on why this style is supported since the `.use` is identical as using `/foo` here (it's specifically testing an edge case for this test).

dougwilson replied:
Ok, so I can see the entire test and it is testing that `app.use('/foo', fn)` runs `fn` for both `/foo` and `/foo/bar`. I think the reason this is tested and I continue to see many, many apps where they use the trailing slash in `app.use` is because it looks similar to `app.use('/', fn)` in that it ends with a slash. I'm not really sure which should be "right", but ideally with the amount of apps this would break, I wonder if it's worth all the turn over in everyone's apps.

Also: I'm describing the issue here correctly, right?

blakeembrey replied:
Sounds correct, though I think you meant `app.use('/foo/', fn)`. With https://github.com/pillarjs/path-to-regexp/issues/123, it'll continue to work, but the child router won't pass in this example since it'll match as `/foo/` or `/` (whatever is input specifically) and the route trimmed will end up as `/foo/bar` -> `bar`. I definitely understand the potential churn here, but I've seen a few issues to the opposite effect which I'm trying to handle here - the unexpected behaviour of `'/'` becoming just `''` means you can't actually mount anything at specifically `/foo/` instead of `/foo`. Previous versions have gotten around the issue by saying "set `strict: true`", and if that's desired I can certainly do that.

wesleytodd replied:
So are we just waiting to publish the `2.1.0` (or whatever you plan to version this as)?  Or is there anything else we need to get this pulled together?

dougwilson replied:
Yea, it sounds like to bump this to at least 2.1.0 (or back to 1.x and other as 2.x). I can merge whichever into the 2.0 branch that is going on.

blakeembrey replied:
@dougwilson How important is this given it's a backward incompatible release of `path-to-regexp` and, I believe, it's a clearer implementation today. Here's a couple of related discussions: https://github.com/pillarjs/path-to-regexp/issues/120 and https://github.com/pillarjs/path-to-regexp/issues/123. With this change, the features in these issues would not be possible (specifically the trailing slash enforcement in https://github.com/pillarjs/path-to-regexp/issues/123, which larger worked by accident in 1.x).

dougwilson replied:
I mean, anything can change, I don't mind as long as we have migration instructions or at least a list of all the changes. Are those differences listed in the PR here already?

Of course Express continues to be num 1 framework for Node because the stability (in API, that is). The more differences the more it will invalidate all books and tutorials out there. It's not a blocker, only somethig to think about.

blakeembrey replied:
Makes sense. I'd love to promote not using `app.use('/', fn)` as mentioned, since it's better overall just to skip the regexp (I believe this module handles it as a special case anyway). The only case it would break is the sub-router if it were trimming based on the matched `/`. However, even that I believe is handled with https://github.com/pillarjs/router/blob/ca2d51c145baa1bb4a3d4b55985b728560c328a1/index.js#L320. Let me do a 2.1 release and merge it into here, chances are it _just works_ because of these two handlers already built into the router that may have been handling 0.1.x edge-cases (for 1.x I had tried to bring that logic into `path-to-regexp`, for 2.x I realised pinning a ton of behaviour to Express.js edge cases might be a worse off idea).

blakeembrey replied:
Done, `2.1.0` is released. When you're ready @wesleytodd, feel free to pull it in and see if it works as expected now. From looking at the module, I believe the edge-cases are all handled at the router level already.

wesleytodd commented:
Ok, updated!  Tests passed locally.

wesleytodd commented:
Ok, so I updated the tests to reflect the breaking change.  Before I write an addition to the change log in the readme, can you all take a look and ensure this is what we all agree we should do?  Specifically `use('/foo/' ...)` will no longer match `/foo` or `/foo/bar`.

You can check these routes with the new behavior of `2.1.0` at https://wesleytodd.github.io/express-route-tester/

dougwilson commented on test/router.js:
```diff
@@ -742,12 +742,12 @@ describe('Router', function () {
       request(server)
       .post('/foo/bar')
-      .expect(200, 'saw POST /bar', cb)
+      .expect(404, cb)
```
I mean, there is no way this is actually the intended behavior, right? If `/foo/` is not a prefix of `/foo/bar` I don't know what `.use` is supposed to be doing.

wesleytodd replied:
Well you you be required to do `.use('/foo')` If you wanted to match `/foo/bar`.

dougwilson replied:
Yea, but then what will it match? If having the trailing / is not correct, then why not throw? I'm afraid this is going to be very confusing.

dougwilson replied:
Using https://wesleytodd.github.io/express-route-tester/ I choose 2.1.0 and it says `/foo/` with no check boxes matches `/foo/bar`. It seems either that test is wrong or there is something else going on here that needs to be fixed.

wesleytodd replied:
`use('/foo/'` will match `/foo/` and nothing else.  It is not that it is "not correct", it just has a very specific use case.  If you wanted the old behavior, all you have to do is change to `use('/foo'` I think.

wesleytodd replied:
You need to have `end` checked to match what is un a use call I think:

```
var layer = new Layer(path, {
  sensitive: this.caseSensitive,
  strict: false,
  end: false
}, fn)
```

dougwilson replied:
If end is checked, that means `end: true`, but like you posted, `end` is false, which is why  unchecked it. The `.use` is the in-ending paths.

wesleytodd replied:
Yeah sorry, I should have stopped before posting that, deleted right after posting it lol.

dougwilson replied:
haha. Cleaned up the convo :)

dougwilson replied:
That defeats the entire purpose of `.use`, does it not? How is `/foo/` not a prefix of `/foo/bar` ? Does `/foo/bar` not begin with `/foo/` ?

wesleytodd replied:
The point of `.use` is maintained without the trailing `/`, and I thought this was reasonable.  To get that behavior back we would have to do make an exception for the behavior of `path-to-regexp`.  AFAIK we only do that for `*` and `/`, adding this exception would be a bit more involved because it is any path with a trailing `/`, and possibly involve other edge cases.

Would it be alright if we directly documented this new behavior?  Or do you think working on this code to add this exception back at the router layer would be good?

dougwilson replied:
I'm not sure. To me, `.use('/foo/', fn)` seems to cause completely unexpected behavior from how Express.js `.use` has always worked forever before. It's in many books and I think breaking this is going to be a large uphill battle from a user support perspective.

I think that either (a) we need to make it work as expected or (b) it shouldn't even be an accepted argument and throw / error out early.

I'm actually curious, perhaps before we proceed, that if you were to document this "new" behavior, what would it sound like / what would the use-case for this type of argument for `.use` be? It just seems really weird if `.use` works two completely different ways depending just on if you happen to add a trailing `/` on the string argument. Seems very unintuitive, but maybe I just don't understand the use-case this new behavior is enabling?

dougwilson replied:
Also, perhaps the answer is that `path-to-regexp` no longer meets the needs of Express.js and we need to scrap it for something else. Just putting things on the table.

wesleytodd replied:
> large uphill battle from a user support perspective

Yeah we should avoid that if reasonable.  FWIW, if I ever wanted to serve a path like this I never put a trailing `/`.  I am not sure how common either one is but every project I have worked on has not used the trailing `/`.



>  (b) it shouldn't even be an accepted argument and throw / error out early.

I think we need some version of the trailing slash if someone wants specifically to do something like serve a directory listing in this case.  So erroring is not really an option.

> I think that either (a) we need to make it work as expected

I am looking at what this would take.  I think it is more simple than I first thought.  See the new commit I added.  Is this satisfactory?

> Also, perhaps the answer is that path-to-regexp no longer meets the needs of Express.js

I would hope we can avoid this unless we actually wanted to change the routing/matching semantics significantly.

wesleytodd commented:
I was looking through the remaining list of things to do for express 5.0, and this is one of the slated changes.  After the last additions does anyone see any issues publishing this and moving forward on `router@2`?

dougwilson commented:
I still think that `.use` is broken with this change, so I would not want to merge it yet.

wesleytodd commented:
Ok, I added an exception so that when we are not in strict mode it will have the existing behavior.  Calls to `.use` are always non strict mode, so `use` works as before.  Does that work?

blakeembrey commented:
Would it help if I identified all the edge cases and the module issued deprecation warnings when encountered? I don’t think any of the cases, except the primary RegExp changes, should be relied upon anyway, but it should be simple to notify the user on how to handle it properly - would be a good opportunity to make sure all these edge cases are legitimate too or whether I’ve removed some useful functionality.

The deprecation warning could be skipped with an environment variable (or enabled with one as an “upgrade guide”).

blakeembrey commented:
Another idea is this module could accept functions as the path argument to easily allow people to switch in other path matching semantics. Upgrading to 2.x is probably in many people’s interests though, I know the latest versions are in use by a lot of front end frameworks that are likely trying to support isomorphic use cases and right now the paths don’t match up with Express.

wesleytodd commented:
> Would it help if I identified all the edge cases and the module issued deprecation warnings when encountered?

I don't think that would help all that much.  That could just as easily be done in express.  I think the issue here is how behind we are, and how difficult it has been to update because it was neglected for so long.  I think in the future this should be solved by updating more regularly.

> Another idea is this module could accept functions as the path argument to easily allow people to switch in other path matching semantics.

IMO the value is having a common set of understood semantics.  Having it be optionally different would mean we could not get the value out of the community agreement on these semantics.  Again I think pushing updates more regularly would be valuable :)

wesleytodd commented:
Just pushed an update to remove the other test changes.  It was failing for array of path conditions, but everything looks good now.  Are we good to merge this?

wesleytodd commented:
I rebased this and pointed it to the `2.0` branch, but noticed that 3.0 is out.  @blakeembrey should I update this to that?  What broke between 2.0 and 3?

blakeembrey commented:
@wesleytodd See https://github.com/pillarjs/path-to-regexp/blob/master/History.md#300--2019-01-13. Specifically it was the treatment around delimiter tokens (e.g. the character before `:token`), it was improved to handle a couple of use-cases more elegantly and removed "partial" token support which is actually better handled by escaping characters.

wesleytodd commented:
Awesome, so I updated to `3.0`, it caused one test failure.  I added another section to the readme to describe the change, and made the corresponding test change.  This should be good to merge now I think.

wesleytodd commented:
Too many versions too fast :P

@blakeembrey Should we update this to `5`? What are the breaking changes in 4 and 5?

blakeembrey commented:
@wesleytodd See https://github.com/pillarjs/path-to-regexp/releases. 4.0 was for ES2015, 5.0 was because I added a feature(s) I shouldn't have.

wesleytodd commented:
Ahh, I was wondering about that quick jump from 3 to 5.  There was too many notifications for me to read all of them and follow.  Could you summarize the changes so we can add it to the readme and docs?  I will update the PR from there.

blakeembrey commented:
@wesleytodd The only mentionable feature is support for nested non-capturing regex groups, e.g. `/(abc(?=d))`. No breaking changes in these releases for paths, only library usage (e.g. `module.exports` -> `module.exports.pathToRegexp`).

blakeembrey commented:
Final release I intend to make for this library is https://github.com/pillarjs/path-to-regexp/releases/tag/v6.0.0, which reverts some of the confusion I added in v3 and adds an alternative (`{}`) to support the reason I did the any character prefix in v3.

blakeembrey commented on README.md:
````diff
@@ -382,6 +382,64 @@ router.route('/pet/:id')
+console.log(req.params) // {'foo': 'bar'}
+```
+
+#### Partial matching, prefer escaping delimiter
````
This section shouldn't be relevant anymore. However, you can make the example `/user{s}?` if you wanted to.

blakeembrey commented on test/route.js:
```diff
@@ -645,7 +645,7 @@ describe('Router', function () {
           var cb = after(2, done)
           var router = new Router()
-          var route = router.route('/user(s)?/:user/:op')
+          var route = router.route('\\/user(s)?/:user/:op')
```
Should be good to revert this.

gireeshpunathil commented:
this needs a rebase. anything else?

dougwilson commented:
This needs nothing at this time as Wes rebased it on a separate branch instead of within the PR. Github will close it automatically after I move the changes back to the PR branch.

dougwilson commented:
So it looks like I don't have permission to push up to the remote branch. This is just a heads-up, so the changes will end up landing on the 2.0 branch, and will end up stating this PR is "closed" instead of "merged".

gireeshpunathil commented:
> the changes will end up landing on the 2.0 branch, and will end up stating this PR is "closed" instead of "merged".

@dougwilson - can you clarify what does this mean? the PR is originally destined for 2.0 branch itself, right? or is there a change? or is it that the landing will be same (to 2.0) but the github web interface will show it as closed, while the code would have landed?

dougwilson commented:
This PR has merge conflicts that must be resolved, and even has an incomplete section in the migration guide among other changes that are needed. I am working on the changes, but I get permission denied trying to push them up to the PR branch. Since I cannot push them into this PR first, landing this PR + changes onto 2.0 will cause GitHub to mark this PR as "closed" instead of "merged", is all.

dougwilson commented:
All the changes here + various fix ups has landed into the `2.0` branch 🎉
